### DISCUSSION TITLE
Rad Pro: Suggestions for new features/ideas

### DISCUSSION THREAD

**Discussion by Davegsm82** [Gissio/radpro#47] [Ideas]
Firstly thanks Gissio for developing this firmware, and secondly I would like to suggest a couple of modifications which may improve usability/safety and/or battery life. 

I've developed my own devices, one of which isn't a million miles away from the overall hardware architecture of the GC-01 but based on a PIC micro and has other stuff like GPS and SD card etc (under the radiacode in the attached picture). 

1 - HV/Tube failure detection - If no pulses from the GM tube are detected within a defined period of time, set an alert (visual/audible) to notify the user there may be an issue. 

2 - A potential power saving feature - the crude nature of the HV generator circuit in the GC-01 is a bit like holding your foot on your cars accelerator and pressing the brake to regulate your speed. It should be possible to gradually increase the PWM pulse width until GM pulses begin to be detected and set that (+a little extra if the HV runs from B+, I didn't check) as a baseline for minimum PWM at ambient radiation levels. Then whenever in elevated radiation levels the PWM could simply be ramped up to the factory default (which appears to be 9.2khz - 75%) to ensure correct operation. 

3 - Serial/USB command/control? - potentially via Geigerlog - more for development purposes than anything else, USB logging is great, would be awesome to be able to send commands to the device to change parameters/UI settings. 

4 - GM Tube self-irradiation/background offset null - Something sadly missing from all cheaper GM Counters, the ability to subtract a small amount of the counts/dose rate/whatever. Basically, everything is radioactive to some degree, including the GM tube and components within the counter, and this can really skew the results in low dose rate environments. With the Default config my GC-01 shows 0.17uSv/h on my desk at work, but I know from my own and company owned and calibrated equipment that the dose rate is actually pretty much 0.05-0.07uSv/h at all times here (I work with LINACs and live sources). In my own device I subtract about 0.08uSv from the reading which brings it surprisingly close to my Radiacode (ignoring dosimetric calibration). 

5 - Tube saturation detection - Unlikely anyone would ever encounter this, but it's a great safety feature. A saturated tube (or a gassy/shorted one) will basically appear like a resistor, so the GM counts input to the microcontroller will see this as a constant high/low signal instead of a pulse (sort of the opposite of #1 above).

6 - Clicker threshold/beeper - for people who have counters running pretty much constantly like myself, it's nice to have a way to gate the clicker so it isn't going constantly and annoying people. Similarly you don't want to have your dose rate alarm set so low that it frightens people around you when you walk past someone who's had TC-99 or you encounter some other elevated radiation level. A nice way of doing this is applying a threshold to the clicker level, so it only clicks above say 0.3uSv for example. 
6.1 - This can be coupled with the display timeout function so that when the screen is on the threshold is disabled, and when it's off then only when above a certain level it will start clicking. 
6.2 - Also, an option for beeps instead of clicks? I implemented a function to change the volume/pitch of beeps depending on radiation level. 

![photo_5771515707803156788_y](https://github.com/Gissio/radpro/assets/88677230/27b94dde-6f93-4c5b-a187-ab30bd46a755)

**Comment by Gissio** [comment]
Thanks for the suggestions. Will be taken into consideration for the next versions.

**Reply by Gissio** [reply]
A lot of your suggestions made it to 2.0rc3.

The second suggestion is a bit of a problem. As you work with LINACs, I can afford a more technical explanation. It is my understanding that the radiation level-count rate curve can be non-monotonic (e.g. https://www.nature.com/articles/s41598-021-81571-3/figures/1 ). Therefore, a given count rate might correspond to a lower or a higher radiation level value. Therefore, it would be dangerous to select the tube voltage solely on the count rate. You might argue that, of course, it is possible to lower or increase the voltage and determine which region of the curve is active. But this requires a lot of experimental effort (Rad Pro supports 7 device variants so far) and time, which I lack.

Regarding your third suggestion. Is it so hard to set up Rad Pro with the keyboard? I don't see a favorable cost-benefit in adding this capability.

Beeps are unfortunately not possible due to hardware limitations.

I'm eager to hear your comments.

**Comment by dc1rdb** [comment]
Thanks Gissio for this great firmware. I have a suggestion for a future version, too.
How about some color coding on the history bargraph, e.g. indicating user defined thresholds like warning and danger thresholds, similar to this screenshot?

![IMG_20240326_161805](https://github.com/Gissio/radpro/assets/9314230/b22565f2-9310-41e0-a268-1c423377d758)

**Reply by Gissio** [reply]
Dankeschön!

**Reply by Gissio** [reply]
Implemented in 2.0rc3 using the [OpenBridge design system](https://www.openbridge.no/)!

**Comment by DG0MG** [comment]
@Davegsm82

I find it very interesting to read your thoughts on RadPro. I have also written a series of suggestions, perhaps you have read them: [https://github.com/Gissio/radpro/discussions/40](https://github.com/Gissio/radpro/discussions/40)
Of course, the suggestions do not only refer to RadPro alone, but may even reach commercial manufacturers who sometimes completely ignore the wishes of the users in the features of their software. Perhaps the reason for this is that the programmer is not a "professional". 

#1 is a great idea and certainly not difficult to implement. If there is no impulse for a minute (or 2 minutes) to display HARDWARE ERROR, it can save lives in certain circumstances.

#2 I had suggested adding a control loop to save power, but that requires a hardware modification. I find it very difficult to do this solely via software, given the large number of supported devices.

#4 Yes, with larger meter tubes (which could be connected) there is an inherent zero effect that must be subtracted from the calibration factor before calculation.

#6 This is the case with the RADEX ONE: If an alarm threshold is set, then the ticker only starts ticking when this is exceeded.

Regards,
 - Steffen, DG0MG -

___
visit [https://www.geigerzaehlerforum.de](https://www.geigerzaehlerforum.de/index.php?action=unread)

**Comment by Wolftech-github** [comment]
First of a big thank you to Gissio for the development of the Rad Pro replacement firmware - amazing work!

I am currently running the v2.0rc2 on a late 2023 model FNIRSI GC-01 that originally came with the HH614 tube, it has since been upgraded with a J305 tube.

The GC-01 communicates via USB-serial with my Mac computer and a small test program shows the device parameters, sets the device time as well as downloads the device log data to a file.

However as noted elsewhere, the firmware upgrade has to be performed on a Windows computer - in my case a borrowed unit, thus i can't easily reflash the device to reset the device log data. This leads me to a couple of suggestions in addition to the ones mentioned above:

Add a menu item/key shortcut in the Data logging menu and/or add a serial command to reset the device log data.

Display the number of device log entries in the Statistics summary.

/Wolf

**Reply by Gissio** [reply]
There is actually a design decision for not adding data log reset: flash wear. Not doing so also keeps the design simpler (on the GC-01 with a WCH, I'm actually close to running out of flash space).

If you wish to download only new data, simply store the timestamp of the most recently downloaded record. Then, use the `GET datalog` command with the `timestamp` parameter, and Rad Pro will stream only the newest records.

I thought about showing datalog statistics in the Statistics view, but it's not that easy. Will be considered for future releases.

**Comment by gruzefix** [comment]
First off, naturally mad props to the dev, you've turned a barely functioning Geiger counter into a powerful device.
I don't know where else to ask this but I feel like this might be the proper place:

Is there any way to sync the device's internal clock with, say, an Android phone connected via USB-C?

The reason I'm asking is because I want to conflate radiation data with GPS data, if possible automatically. I want to go to different places and see if there's radiation hotspots in certain ones. I live near an area that a) has been affected by Chernobyl fallout more than other surrounding areas and b) used to be a swamp that has only been drained in the past 2-300 years, so there's still lots of radon coming out of the ground. I used to spend a significant part of my clinical rotation in oncology and a few of my colleagues and I noticed that unusually many of our patients seem to come from [that specific area](https://en.wikipedia.org/wiki/Old_Bavarian_Donaumoos). The area is pretty flat, so it's ideal for a Sunday bike tour in the summer, so why not gather data while you're at it?

Or are there already any readily available solutions? GMCMap seems promising but from what I gather it only features stationary detectors, not mobile ones.

--gruzefix

**Reply by Gissio** [reply]
`radpro-tool` does that (from a PC): https://github.com/Gissio/radpro/blob/main/docs/manual.md#radpro-tool

The clock skew within a day should be sufficient for your requirements.

May I also suggest that you contact some government authority, radiology institute or university department before starting your undertaking? I am sure they will be able to give you good advice.

**Reply by Gissio** [reply]
@gruzefix: it my understanding that you are a native Ukrainian speaker. Could you check the Ukrainian translation in 2.2beta3?

The strings are available here: https://github.com/Gissio/radpro/blob/main/platform.io/src/strings/uk.h

If you are willing to help us with these, please follow the discussion here: https://github.com/Gissio/radpro/discussions/241

**Reply by SunboX** [reply]
@gruzefix Hi

> Is there any way to sync the device's internal clock with, say, an Android phone connected via USB-C?

I'm able to build such a thing. Is this still from interest? Anything else that should go in such a app? Import / Export formats, logging and so on? Anything that you would specifically need? Or any specific format? Like some kind of protocol PDF export or so?

greetings,
André

**Comment by Thijzewijs** [comment]
Having upgraded the Fnirsi GC-01 with the 3.0.2 Radpro alternative I'm very impressed! I do have another idea that might be possible to implement: keep the screen on when powered externally. When I have the unit on USB power, it continuously charges but it would also be nice to have the screen on (configurable in the settings menu). The current draw is about 400-500mA while charging so about every USB powersource can handle the extra current needed to power the screen.

Keep up the good work!

**Reply by Gissio** [reply]
Set **Display > Sleep > Always on**.

**Reply by Thijzewijs** [reply]
True but that's turning on the display, regardless of charging or not. An extra function that enables always on when the charger is connected which returns to the already available setting when the charger is disconnected.

In that case the display has a time out like today, for instance 30 seconds, and when the charger gets connected the time out is overruled and the display stays on.

Hopefully that clarifies what I meant to say.

**Comment by zolygit** [comment]
Thank you Gissio for the Firmware. I've just swapped out the original on a Bosean 5000. Much better.
Just a quick question. 
Would it possible to implement 3 - 4 point calibration for the tubes?
I believe GMC firmware's has a 2 point one.
We really appreciate what you are doing! 
Thank you.

**Comment by Gissio** [comment]
@zolygit: thanks for your comments and suggestions.

Here are my thoughts:

* No professional radiation expert would rely on the non-linear response region of a standard Geiger tube for high dose rates. Instead, they use specialized tubes designed for those conditions, such as the SI3BG or LND 713. Guidelines from bodies like the IAEA and NRC emphasize using appropriate detectors to ensure measurements stay linear.
* Rad Pro is aimed at the general public, so features like dead-time compensation might seem overly complex (they involve concepts, tools and procedures that aren't immediately obvious). Yet, the firmware includes a practical way for users to measure an upper-bound estimate of the tube's dead time and apply a simple correction. Even with some experimental uncertainty, applying this basic correction is far better than leaving the readings uncorrected at higher count rates.
* I’m aware that some hobbyists are interested in multi-point calibration, which can be useful in certain niche scenarios. However, I see it as too specialized for the core goals of this project.

For these reasons, multi-point calibration won't be added to Rad Pro. That said, the project is open-source—you're more than welcome to fork it and implement the feature yourself if it fits your needs.

**Reply by zolygit** [reply]
Thank you for the detailed explanation.

**Comment by multicoolaz** [comment]
Greetings from a new happy Bosean5000+RadPro user. Thank you for what you are doing! 

Just created an account to suggest this: a "Fridge Monitoring" mode for Rad Pro.

The idea is to use the detector for long-term monitoring of a refrigerator and detect a small increase in radiation caused by newly added food products, i.e. cesium-contaminated blueberries.

The mode could work similarly to "Timed Dose", but with the following changes:

The user sets a measurement interval, from 30 minutes to 24 hours. Optional: scheduled cycle start.
When the interval ends, the measurement automatically restarts from zero, creating a continuous cyclic monitoring process.
The detector continuously accumulates the actual dose during this interval.
An alarm threshold can then be configured as an excess dose above the baseline (which is measured in an empty fridge).

The alarm may have separate settings for visual alarm only (LED/display) or audible + visual alarm (no vibro).
The visual-only mode is important because the detector may be left unattended for many hours or overnight.
After an alarm, the monitoring cycle may either stop and display the dose, or continue cycling while logging the results and flashing red for the user to notice upon return, so the device can remain unattended.

The key point is that the alarm should be based on the accumulated dose over a long period, rather than simply comparing the instantaneous CPM or average count rate. This makes it possible to detect very small but persistent increases in radiation continuously and automatically. In case of an alarm, the user can investigate it with more specialized devices, such as a beta detector or a spectrometer.

A shorter monitoring interval would provide faster detection of a larger increase, while a longer interval would allow the system to detect a much smaller increase that would be difficult to distinguish from normal statistical fluctuations over a short measurement period.

This feature could be useful not only for refrigerators, but also as a general long-term radiation monitoring mode for unattended locations or turnover storage areas.

**Comment by Gissio** [comment]
Thanks, and welcome.

Unfortunately the physics is not on your side. A kilo of berries at the EU limit (600 Bq/kg Cs-137) adds roughly 0.8 cpm to the FS-5000's ~17 cpm background, a 5% effect. Counting statistics could reach that in a few hours, but other things can as well: ordinary clean groceries carry comparable K-40 activity (30–150 Bq/kg), indoor radon swings tens of percent with weather and ventilation, and the tube's HV drifts at 4 °C. Your alarm would fire after every big shop. Also, a bare GM tube can't tell you what it's counting: the screening step needs a spectrometer too.

**Comment by DG0MG** [comment]
It is worth noting, however, that this very tiny increase in the count rate occurs only when the measuring device is in direct contact with the material being measured (i.e., the berries). At a distance of 30 or 50 cm—for instance, if the FS-5000 were simply lying somewhere in the refrigerator—nothing would be detected given the levels of contamination encountered in 'normal' real-world scenarios.
I know that Chinese advertising often portrays a scenario where you briefly hold one of their cheap radiation detectors against a product in the supermarket and supposedly know whether or not to buy it. But that is more wishful thinking and fantasy than reality.
Please take a look at how food is actually measured: using a highly sensitive detector (such as a large-area Geiger tube or a big scintillator) housed within thick lead shielding to eliminate background radiation, and employing a Marinelli geometry—where the detector is surrounded on all sides by the sample. Ideally, a spectrometer is used to identify the potassium-40 content and distinguish it from the cesium-137 content.

**Comment by multicoolaz** [comment]
Thanks for the detailed critique — the K-40 and geometry points are both important. To mitigate them, I could keep potatoes out of the same bin as the main cesium accumulators — berries and mushrooms — and instead place the detector directly in the bin with them. Radon is also essentially removed from my kitchen already.

The 0.8 cpm/kg figure for berries is indeed quite small. But what if the trigger case were mushrooms instead? Cesium "accumulators" (like Polish bolete, slippery jacks, or Lactarius species) in post-Chernobyl forests show activity an order of magnitude above the berry limit — thousands of Bq/kg versus 600. At those levels the effect is no longer 5%, but tens of percent above background, and K-40 from the mushroom itself doesn't mask it. For this case, the accumulation mode still looks statistically justified.

Fully surrounding the sample (quasi-Marinelli geometry) would give a 1.5-2x gain in efficiency. A second reference detector outside the mushroom bin (on another shelf of the fridge, for instance), instead of a static baseline, could compensate for drift from radon, HV, and temperature: the real-time difference between the two channels would be compared instead.

On "you need a spectrometer" — I fully agree, which is why the original proposal frames the mode as a trigger for further investigation, not a final diagnosis. Incidentally, the RadiaScan-701, with its pancake detector already has a commercial "Detection" mode built on exactly this background-vs-object logic for food products — it's just designed for a single short measurement. I'm essentially proposing the same concept, stretched out over time and with less maintenance — and not necessarily limited to fridges.

Such a tool could also be used in high-turnover storage, like parcel pickup points. The main goal is a watchdog that's more sensitive thanks to accumulating dose over a set period, without requiring constant manual monitoring. A feature like this would add value to a device built on a cheap tube, though for my own fridge I'll probably end up with a scintillator sooner or later. Thanks again for your responses.
